query-planner-js/src/composedSchema/buildComposedSchema.ts

```diff
@@ -0,0 +1,240 @@
+  // to handle renames of @core itself.
+  const coreName = 'core';
+
+  const coreDirective = schema.getDirective(coreName);
+
+  assert(coreDirective, `Expected core schema, but can't find @core directive`);
+
+  // TODO: We should follow the CollectFeatures algorithm from the Core Schema
+  // spec here, and use the collected features to validate feature
+  // versions and handle renames.
+
+  const coreDirectivesArgs = getArgumentValuesForRepeatableDirective(
+    coreDirective,
+    schema.astNode!,
+  );
+
+  for (const coreDirectiveArgs of coreDirectivesArgs) {
+    const feature: string = coreDirectiveArgs['feature'];
+
+    if (
+      !(
+        feature === 'https://lib.apollo.dev/core/v0.1' ||
+        feature === 'https://lib.apollo.dev/join/v0.1'
+      )
+    ) {
+      throw new GraphQLError(
+        `Unsupported core schema feature and/or version: ${feature}`,
+        schema.astNode!,
+      );
+    }
+  }
+
+  const joinName = 'join';
+
+  function getJoinDirective(name: string) {
+    const fullyQualifiedName = `${joinName}__${name}`;
+
+    const directive = schema.getDirective(fullyQualifiedName);
+    assert(
+      directive,
+      `Composed schema should define @${fullyQualifiedName} directive`,
+    );
+    return directive;
+  }
+
+  const ownerDirective = getJoinDirective('owner');
+  const typeDirective = getJoinDirective('type');
+  const fieldDirective = getJoinDirective('field');
+  const graphDirective = getJoinDirective('graph');
+
+  const graphEnumType = schema.getType(`${joinName}__Graph`);
+  assert(isEnumType(graphEnumType));
+
+  const graphMap: GraphMap = Object.create(null);
+
+  schema.extensions = {
+    ...schema.extensions,
+    federation: {
+      graphs: graphMap,
+    },
+  };
+
+  for (const graphValue of graphEnumType.getValues()) {
+    const name = graphValue.name;
+
+    const graphDirectiveArgs = getArgumentValuesForDirective(
+      graphDirective,
+      graphValue.astNode!,
+    );
+    assert(
+      graphDirectiveArgs,
+      `${graphEnumType.name} value ${name} in composed schema should have a @${graphDirective.name} directive`,
+    );
+
+    const serviceName: string = graphDirectiveArgs['name'];
+    const url: string = graphDirectiveArgs['url'];
+
+    graphMap[name] = {
+      serviceName,
+      url,
+    };
+  }
+
+  for (const type of Object.values(schema.getTypeMap())) {
+    if (isIntrospectionType(type)) continue;
+
+    // We currently only allow join spec directives on object types.
+    if (!isObjectType(type)) continue;
+
+    assert(
+      type.astNode,
+      `GraphQL type "${type.name}" should contain AST nodes`,
+    );
+
+    const ownerDirectiveArgs = getArgumentValuesForDirective(
+      ownerDirective,
+      type.astNode,
+    );
+
+    const typeMetadata: FederationTypeMetadata = ownerDirectiveArgs
+      ? {
+          serviceName: graphMap[ownerDirectiveArgs?.['graph']].serviceName,
```

Comment:
This code assumes that the `graph` arg value is found in graphMap, which seems reasonable because it's an enum value and we've already set up graphMap with all enum values. But this relies on (a) the assumption that the `@join__owner` directive is declare with `graph: join__Graph!`, which we have not validated, and (b) the assumption that graphql-js validates directive argument types which... um, I think maybe `getArgumentValues` does even if [schema parsing does not](https://github.com/graphql/graphql-js/issues/1389#issuecomment-416929486).

OK but (a) still matters: let's make sure we get a better error message if the graph isn't in graphMap!

Part of this is a problem with the type

```
export type GraphMap = { [graphName: string]: Endpoint };
```

This type means "no matter what string you put in, you'll get an Endpoint out", which is not true. The type should be changed to `Endpoint | undefined`, or we can switch to an actual `Map` which handles this better.

I guess you could also leave this GraphMap type and validate both `@join__*` directives have the right argument. Or make both changes.